File: rust/src/day3.rs
use std::collections::HashSet;

use std::str::FromStr;

use crate::util::load;

struct Rucksack {
    comp1: HashSet<char>,
    comp2: HashSet<char>,
}

impl FromStr for Rucksack {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut comp1 = HashSet::new();
        let mut comp2 = HashSet::new();
        for (i, c) in s.chars().enumerate() {
            if i < s.len() / 2 {
                comp1.insert(c);
            } else {
                comp2.insert(c);
            }
        }
        Ok(Rucksack { comp1, comp2 })
    }
}

pub fn priority(c: &char) -> u32 {
    match c {
        'A'..='Z' => Into::<u32>::into(*c) - Into::<u32>::into('A') + 27,
        'a'..='z' => Into::<u32>::into(*c) - Into::<u32>::into('a') + 1,
        _ => 0,
    }
}

pub fn part1() -> u32 {
    let rucksacks: Vec<Rucksack> = load("data/day3.txt");
    rucksacks
        .into_iter()
        .map(|r| {
            r.comp1
                .intersection(&r.comp2)
                .map(|c| priority(c))
                .sum::<u32>()
        })
        .sum()
}

pub fn part2() -> u32 {
    let rucksacks: Vec<Rucksack> = load("data/day3.txt");
    rucksacks
        .chunks(3)
        .map(|chunk| {
            chunk
                .into_iter()
                .map(|r| r.comp1.union(&r.comp2).copied().collect())
                .reduce(|acc: HashSet<char>, r| acc.intersection(&r).copied().collect())
                .unwrap()
                .into_iter()
                .map(|c| priority(&c))
                .sum::<u32>()
        })
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part1() {
        let priority_sum = part1();
        println!("Priority sum is {}", priority_sum);
        assert_eq!(priority_sum, 7674);
    }

    #[test]
    fn test_part2() {
        let priority_sum = part2();
        println!("Priority sum is {}", priority_sum);
        assert_eq!(priority_sum, 2805);
    }
}

File: python/day3.py
from functools import reduce
from typing import Set, List

from util import load


def chunk(rucksacks: List[str], n: int) -> List[List[Set[str]]]:
    chunks = []
    for i in range(0, len(rucksacks), n):
        chunks.append([set(r) for r in rucksacks[i:i + 3]])
    return chunks


def split_in_halves(rucksack) -> List[Set[str]]:
    half = len(rucksack) // 2
    return [set(rucksack[:half]), set(rucksack[half:])]


def common_part(*args) -> str:
    return reduce(lambda s1, s2: s1.intersection(s2), *args).pop()


def priority(item: str) -> int:
    if item.isupper():
        return 27 + ord(item) - ord('A')
    else:
        return 1 + ord(item) - ord('a')


def part1(rucksacks: List[str]) -> int:
    return sum(
        map(priority,
            map(common_part,
                map(split_in_halves, rucksacks))))


def part2(rucksacks: List[str]) -> int:
    return sum(
        map(priority,
            map(common_part,
                chunk(rucksacks, 3))))


def main():
    rucksacks = load('day3.txt')

    sum1 = part1(rucksacks)
    print(f'Part 1: {sum1}')
    assert sum1 == 7674

    sum2 = part2(rucksacks)
    print(f'Part 2: {sum2}')
    assert sum2 == 2805


if __name__ == "__main__":
    main()
